keras/engine/training.py

```diff
@@ -1403,3 +1403,65 @@ def evaluate_generator(self, generator, val_samples):
                 averages.append(np.average([out[i] for out in all_outs],
                                 weights=weights))
             return averages
+
+    def predict_generator(self, generator, val_samples):
+        '''Generate predictions for the input samples from a data generator.
+        The generator should return the same kind of data as accepted by
+        `predict_on_batch`.
+
+        Arguments:
+            generator:
+                generator yielding input samples
+            val_samples:
+                total number of samples to generate from `generator`
+                before returning.
+
+        # Returns
+            A Numpy array of predictions.
+        '''
+        processed_samples = 0
+        wait_time = 0.05
+        all_outs = []
+        data_gen_queue, _stop = generator_queue(generator)
+
+        while processed_samples < val_samples:
+            generator_output = None
+            while not _stop.is_set():
+                if not data_gen_queue.empty():
+                    generator_output = data_gen_queue.get()
+                    break
+                else:
+                    time.sleep(wait_time)
+
+            if isinstance(generator_output, tuple):
+                if len(generator_output) == 2:
+                    x, y = generator_output
+                    sample_weight = None
+                elif len(generator_output) == 3:
+                    x, y, sample_weight = generator_output
+                else:
+                    _stop.set()
+                    raise Exception('output of generator should be a tuple '
+                                    '(x, y, sample_weight) '
+                                    'or (x, y). Found: ' + str(generator_output))
+            else:
+                x = generator_output
+
+            x = standardize_input_data(x, self.input_names, self.internal_input_shapes)
+
+            try:
+                outs = self.predict_on_batch(x)
+            except Exception as e:
+                _stop.set()
+                raise e
+
+            nb_samples = len(x[0])
+            if len(all_outs) == 0:
+                all_outs = np.zeros((val_samples,) + outs.shape[1:])
```

Comment:
`all_outs` is meant to be a list of arrays (potentially with 1 element), not a Numpy array, because we need to support multi-output models (the present code wouldn't). Same for `outs`.

So I would recommend following the pattern from `evaluate_generator`: converting the output of `self.predict_on_batch` to a list if necessary, etc.